core/mm/core_mmu.c

```diff
@@ -1193,15 +1193,13 @@ static void assign_mem_granularity(struct memory_map *mem_map)
 	for  (n = 0; n < mem_map->count; n++) {
 		paddr_t mask = mem_map->map[n].pa | mem_map->map[n].size;
 
-		if (!(mask & CORE_MMU_PGDIR_MASK))
-			mem_map->map[n].region_size = CORE_MMU_PGDIR_SIZE;
-		else if (!(mask & SMALL_PAGE_MASK))
-			mem_map->map[n].region_size = SMALL_PAGE_SIZE;
-		else
+		if (mask & SMALL_PAGE_MASK)
 			panic("Impossible memory alignment");
 
 		if (map_is_tee_ram(mem_map->map + n))
 			mem_map->map[n].region_size = SMALL_PAGE_SIZE;
+		else
+			mem_map->map[n].region_size = CORE_MMU_PGDIR_SIZE;
```

Comment:
Previous implementation allowed to page map cacheable memory outside TEE RAM, (MEM_AREA_RAM_NSEC, MEM_AREA_RAM_SEC, MEM_AREA_ROM_SEC). I fear memory firewall permission violation on speculative accesses  for such memories, and likely few other corner issues on some platform specific memories.

Reply:
This is the requested granularity. It will not mean mapping more memory than specified. So some ranges will be mapped by of mix of small-page and pgdir sized mappings (level 3 and level 2 with LPAE).

Take QEMU with hafnium as an example:
```
TEE_RAM_RWX  va 0x0e304000..0x0e503fff pa 0x0e304000..0x0e503fff size 0x00200000 (smallpg)
RES_VASPACE  va 0x0e600000..0x0effffff pa 0x00000000..0x009fffff size 0x00a00000 (pgdir)
SHM_VASPACE  va 0x0f000000..0x10ffffff pa 0x00000000..0x01ffffff size 0x02000000 (pgdir)
IO_SEC       va 0x11000000..0x1a080fff pa 0x09000000..0x12080fff size 0x09081000 (pgdir)
SEC_RAM_OVERALL va 0x1a300000..0x1affffff pa 0x0e300000..0x0effffff size 0x00d00000 (pgdir)
```
`IO_SEC` is then mapped as:
```
D/TC:0        [LVL2] VA:0x0011000000 PA:0x0009000000 DEV-RW-XN-S
D/TC:0        [LVL2] VA:0x0011200000 PA:0x0009200000 DEV-RW-XN-S
...
D/TC:0        [LVL2] VA:0x001a000000 TBL:0x000e3fa000 
D/TC:0          [LVL3] VA:0x001a000000 PA:0x0012000000 DEV-RW-XN-S
D/TC:0          [LVL3] VA:0x001a001000 PA:0x0012001000 DEV-RW-XN-S
...
D/TC:0          [LVL3] VA:0x001a080000 PA:0x0012080000 DEV-RW-XN-S
D/TC:0          [LVL3] VA:0x001a081000 INVALID
```
`SEC_RAM_OVERALL` as:
```
D/TC:0          [LVL3] VA:0x001a300000 PA:0x000e300000 MEM-RW-XN-S
D/TC:0          [LVL3] VA:0x001a301000 PA:0x000e301000 MEM-RW-XN-S
...
D/TC:0          [LVL3] VA:0x001a3ff000 PA:0x000e3ff000 MEM-RW-XN-S
D/TC:0        [LVL2] VA:0x001a400000 PA:0x000e400000 MEM-RW-XN-S
...
D/TC:0        [LVL2] VA:0x001ae00000 PA:0x000ee00000 MEM-RW-XN-S
D/TC:0        [LVL2] VA:0x001b000000 INVALID
```

So pgdir-sized `region_size` uses as few translation tables as possible, but it will not result in mapping beyond the requested range.